src/ray/core_worker/transport/direct_task_transport.cc

```diff
@@ -118,15 +122,205 @@ void CoreWorkerDirectTaskSubmitter::AddWorkerLeaseClient(
+
+  // Double check to make sure that we didn't pick the thief as the designated victim
+  RAY_CHECK(!(victim_addr == thief_addr) &&
+            victim_addr.worker_id != thief_addr.worker_id);
+
+  auto &victim_entry = worker_to_lease_entry_[victim_addr];
+  // Double check that the victim has the correct SchedulingKey
+  RAY_CHECK(victim_entry.scheduling_key == scheduling_key);
+
+  RAY_LOG(DEBUG) << "Victim is worker " << victim_addr.worker_id << " and has "
+                 << victim_entry.tasks_in_flight << " tasks in flight, "
+                 << " among which we estimate that "
+                 << victim_entry.EstimateTasksAvailableForSteal()
+                 << " are available for stealing";
+  RAY_CHECK(victim_entry.tasks_in_flight >=
+            victim_entry.EstimateTasksAvailableForSteal());
+  RAY_CHECK(scheduling_key_entry.total_tasks_in_flight >= victim_entry.tasks_in_flight);
+
+  if (victim_entry.EstimateTasksAvailableForSteal() < 1) {
+    RAY_LOG(DEBUG) << "The designated victim does not have enough tasks to steal.";
+    return false;
+  }
+
+  return true;
+}
+
+void CoreWorkerDirectTaskSubmitter::StealTasksIfNeeded(
+    const rpc::WorkerAddress &thief_addr, bool was_error,
+    const SchedulingKey &scheduling_key,
+    const google::protobuf::RepeatedPtrField<rpc::ResourceMapEntry> &assigned_resources) {
+  // Check if work stealing is enabled
+  if (max_tasks_in_flight_per_worker_ == 1 || !work_stealing_) {
+    RAY_LOG(DEBUG) << "Work stealing is not enabled, so we return the worker "
+                   << thief_addr.worker_id << " without stealing";
+    ReturnWorker(thief_addr, was_error, scheduling_key);
+    return;
+  }
+
+  RAY_LOG(DEBUG) << "Beginning to steal work now! Thief is worker: "
+                 << thief_addr.worker_id;
+
+  auto &thief_entry = worker_to_lease_entry_[thief_addr];
+  // Check that the thief still retains its lease_client, and it has no tasks in flights
+  RAY_CHECK(thief_entry.lease_client);
+  RAY_CHECK(thief_entry.tasks_in_flight == 0);
+  RAY_CHECK(thief_entry.WorkerIsStealing() == false);
+  RAY_CHECK(thief_entry.EstimateTasksAvailableForSteal() == 0);
+
+  // Search for a suitable victim
+  rpc::Address victim_raw_addr;
+  if (!FindOptimalVictimForStealing(scheduling_key, thief_addr, &victim_raw_addr)) {
+    RAY_LOG(DEBUG) << "Could not find a suitable victim for stealing! Returning worker "
+                   << thief_addr.worker_id;
+    ReturnWorker(thief_addr, was_error, scheduling_key);
+    return;
+  }
+  rpc::WorkerAddress victim_addr = rpc::WorkerAddress(victim_raw_addr);
+
+  // Record the new StealTasks request in the victim's LeaseEntry to facilitate the
+  // estimate of how many tasks are stealable
+  RAY_CHECK(worker_to_lease_entry_.find(victim_addr) != worker_to_lease_entry_.end());
+  auto &victim_entry = worker_to_lease_entry_[victim_addr];
+  victim_entry.SetNewStealTaskRequestPending();
```

Comment:
Have you tried running this without setting the current estimate? I'm wondering if it's necessary since it would be nice to estimate steals just based on the number of in-flight tasks. That seems simpler than trying to keep this amount up-to-date, even if it means that there are some failed attempts to steal.

Reply:
I just changed that in commit: https://github.com/ray-project/ray/pull/13570/commits/907b95406b94c252ee4800aa028a1a27e25faa2e and the performance seems to be unchanged